// storybook-addon-aem/dist/esm/decorators/withResources.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withResources = void 0;

var _react = _interopRequireDefault(require("react"));

var _addons = require("@storybook/addons");

var _aemReactEditableComponents = require("@adobe/aem-react-editable-components");

var _Container = require("../components/Container/Container.component");

var _Htl = require("../components/Htl/Htl.component");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var withResources = function withResources(StoryFn, context) {
  /**
   * Global parameters.
   */
  var parameters = (0, _addons.useParameter)("aem");
  /**
   * Register an AEM resource.
   */

  var registerResource = function registerResource(path, render) {
    var MappedComponent = function MappedComponent(props) {
      if (!render) {
        return /*#__PURE__*/_react["default"].createElement("div", null, "No render function specified");
      }

      return /*#__PURE__*/_react["default"].createElement(_Htl.Htl, _extends({
        render: render
      }, props.model));
    };

    return (0, _aemReactEditableComponents.MapTo)(path)(MappedComponent);
  };
  /**
   * Register a list of AEM resources.
   */


  var registerResources = function registerResources(resources) {
    return resources.map(function (resource) {
      var path = resource.path,
          render = resource.render;
      return registerResource(path, render);
    });
  };
  /**
   * Register resources.
   */


  (0, _addons.useEffect)(function () {
    var _parameters$config;

    if (parameters !== null && parameters !== void 0 && (_parameters$config = parameters.config) !== null && _parameters$config !== void 0 && _parameters$config.appId) {
      var _parameters$config2;

      (0, _aemReactEditableComponents.MapTo)("".concat(parameters === null || parameters === void 0 ? void 0 : (_parameters$config2 = parameters.config) === null || _parameters$config2 === void 0 ? void 0 : _parameters$config2.appId, "/components/container"))(_Container.Container);
    }

    if (parameters !== null && parameters !== void 0 && parameters.config.resources) {
      registerResources(parameters.config.resources);
    }
  }, [parameters]);
  return StoryFn();
};

exports.withResources = withResources;
